### DISCUSSION TITLE
Native CSS support opportunities

### DISCUSSION THREAD

**Discussion by barak007** [webpack/webpack#15160] [Ideas]
Hello webpack team.

I am one of the developers of a library named [Stylable](https://stylable.io/), it's a CSS pre-processor which involves a deep [integration ](https://github.com/wix/stylable/tree/master/packages/webpack-plugin) with `webpack` and `mini-css-extract-plugin`.
Currently part of our integration feels a bit hacky because some of the APIs we required are not documented and so we learned from reading the code. 
We saw the initial work around Native CSS support and we think this will be a great opportunity to simplify our integration and use official API's

Due to the complexity of our integration, we'd like to propose a few APIs we'll be needing from the upcoming built-in CSS integration:

* Add/inject additional `CssModule`s from our plugin
* Generate JS modules similar to `CSS modules` but with runtime functionality backed in, and having different export structure
* Exclude `.st.css` file extension from the built-in webpack CSS process or somehow use the built-in integrations as a post-process for asset handling and chunking 
* Optimize the JS modules and emitted CSS respectively before emitting bundles. (e.g. make classnames shorter and remove unused components CSS)

I hope we can start a discussion around adding these APIs/features so we'll be able to continue maintaining a clean and efficient integration to webpack.
Thanks in advance.

**Comment by sokra** [comment]
I think we probably have you covered, but it's still in an early state.

> * Add/inject additional `CssModule`s from our plugin

In webpack you usually don't "generate" modules, but instead write or generate imports that lead to the creating of a module.

> * Generate JS modules similar to `CSS modules` but with runtime functionality backed in, and having different export structure

This sounds like you can create a JS module that wraps a css module.

> * Exclude `.st.css` file extension from the built-in webpack CSS process or somehow use the built-in integrations as a post-process for asset handling and chunking 

Webpack allows to configure the css support via rules.

> * Optimize the JS modules and emitted CSS respectively before emitting bundles. (e.g. make classnames shorter and remove unused components CSS)

The current experimental css modules implementation in webpack does mangle class names. Removing unused rules is planned but not yet implemented.
